stdlib/public/core/Range.swift

```diff
@@ -231,7 +231,14 @@ where Bound: Strideable, Bound.Stride: SignedInteger
   @inlinable
   @inline(__always)
   public func index(after i: Index) -> Index {
-    _failEarlyRangeCheck(i, bounds: startIndex..<endIndex)
+    _precondition(i >= lowerBound)
+
+    // Explicitly don't use 'i < upperBound' because it blocks some
+    // optimizations from occuring for simple loops like '0 ..< n'. This branch
+    // should almost never be taken because it's illegal to craft a Range whose
+    // upperbound is greater than an integer max, so incrementing the index, as
+    // long as we're not the endIndex, will always succeed.
+    _precondition(i != upperBound)
```

Comment:
Does this open up a safety hole for situations where arbitrary indices are easy to construct by hand (e.g. integers)?